File: 2025-04-02/Reorder_List.java
package amazon;

public class Reorder_List {
    static public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public static void reorderList(ListNode head) {
        if(head == null) return;

        // find mid
        ListNode slow = head, fast = head;
        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        // reverse second half
        ListNode prev = null;
        while(slow.next != null) {
            ListNode temp = slow.next;
            slow.next = prev;
            prev = slow;
            slow = temp;
        }
        slow.next = prev;

        // Reorder
        ListNode n1 = head, n2 = slow;
        while(n2.next != null) {
            ListNode temp1 = n1.next, temp2 = n2.next;
            n1.next = n2;
            n2.next = temp1;
            n1 = temp1;
            n2 = temp2;
        }
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        reorderList(head);

        while(head != null) {
            System.out.println(head.val);
            head=head.next;
        }
    }
}

File: 2026-02-02/Reorder_List.py
from typing import Optional
from Linked_List import Linked_List, ListNode


class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head:
            return

        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        cur = slow.next
        prev = slow.next = None

        while cur:
            temp = cur.next
            cur.next = prev
            prev = cur
            cur = temp

        first, second = head, prev
        while second:
            temp1, temp2 = first.next, second.next
            first.next = second
            second.next = temp1
            first, second = temp1, temp2

        return


sol = Solution()
ll = Linked_List()
head = ll.build_from_list([1,2,3,4,])
sol.reorderList(head)
ll.print_list(head)